api/s2n.h

```diff
@@ -596,6 +596,258 @@ extern int s2n_cert_get_utf8_string_from_extension_data_length(const uint8_t *ex
+ * prior to this API call to determine if an offered pre-shared key exists next in the list. 
+ * 
+ * @param psk_list A pointer to the offered pre-shared key list being read.
+ * @param psk A pointer to the next offered pre-shared key object being obtained.
+ */
+S2N_API 
+int s2n_offered_psk_list_next(struct s2n_offered_psk_list *psk_list, struct s2n_offered_psk *psk);
+
+/**
+ * Resets the offered pre-shared key list pointer to the initial pre-shared key object in the order 
+ * in which the pre-shared key objects were originally added to the list by the client.
+ *  
+ * @param psk_list A pointer to the offered pre-shared key list being reread.
+ */
+S2N_API 
+int s2n_offered_psk_list_reread(struct s2n_offered_psk_list *psk_list);
+
+/**
+ * Chooses a pre-shared key from the offered pre-shared key list to be used for the connection.  
+ * This API matches the pre-shared key identity received from the client against the server's known pre-shared key identities 
+ * list, in order to choose the pre-shared key to be used for the connection. If the pre-shared key identity sent from the client 
+ * is NULL, no pre-shared key is chosen for the connection. If the client offered pre-shared key identity has no matching PSK identity 
+ * with the server, an error will be returned. Use this API along with the `s2n_psk_selection_callback` callback to select a PSK identity.
+ * 
+ * @param psk_list A pointer to the server's known pre-shared key list used to compare for a matching pre-shared key with the client.
+ * @param psk A pointer to the client's pre-shared key object used to compare with the server's known PSK identities.
+ */
+S2N_API int s2n_offered_psk_list_choose_psk(struct s2n_offered_psk_list *psk_list, struct s2n_offered_psk *psk);
+
+/**
+ * Callback function to select a pre-shared key from a list of offered pre-shared keys.
+ * Use this callback when a custom PSK selection logic is desired. The s2n-tls default PSK selection logic 
+ * chooses the first matching pre-shared key from the list of offered pre-shared keys shared by the client.
+ * 
+ * # Safety
+ *
+ * `context` is a void pointer and the caller is responsible for ensuring it is cast to the correct type.
+ *
+ * @param conn A pointer to the s2n_connection object.
+ * @param context A pointer to a context for the caller to pass state to the callback, if needed.
+ * @param psk_list A pointer to the offered pre-shared key list being read.
+ */
+typedef int (*s2n_psk_selection_callback)(struct s2n_connection *conn, void *context,
+                                          struct s2n_offered_psk_list *psk_list);
+
+/**
+ * Sets the callback to select the matching pre-shared key (PSK). 
+ * If this callback is not set s2n-tls uses a default PSK selection logic that chooses the first matching 
+ * pre-shared key from the list of offered pre-shared keys shared by the client. 
```

Comment:
This isn't actually the default logic. We use server preference order, not client preference.

Reply:
Right that is what I was trying to convey here - Does adding the `server` term remove the confusion for you: 
 `chooses the first matching  server pre-shared key from the list..`

Reply:
Idk I think that still doesn't communicate the idea clearly enough. The problem might be that this information belongs in the more general "how to use PSK" usage guide, or even on the append API instead, since it's more about the meaning of PSK order than the callback itself.

Maybe:
If this callback is not set, the server chooses the first identity in its pre-shared key list that also appears in the client's  pre-shared key list.